TS error when specifying schema.pre 'deleteOne' with { document: true }

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

7.3.2

### Node.js version

16.18.1

### MongoDB server version

4.4

### Typescript version (if applicable)

5.1

### Description

So basicallly specifying `schema.pre('deleteOne', { document: true }, function() {});` will throw an error:
```
error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '"deleteOne"' is not assignable to parameter of type 'RegExp | "insertMany"'.

168  schema.pre('deleteOne', { document: true }, async function preHandler() {
```

Runtime-wise everything works though.

### Steps to Reproduce

```
schema.pre('deleteOne', { document: true }, function() {});
```

### Expected Behavior

_No response_

--- a/test/types/middleware.test.ts
+++ b/test/types/middleware.test.ts
@@ -1,5 +1,5 @@
 import { Schema, model, Model, Document, SaveOptions, Query, Aggregate, HydratedDocument, PreSaveMiddlewareFunction } from 'mongoose';
-import { expectError, expectType, expectNotType } from 'tsd';
+import { expectError, expectType, expectNotType, expectAssignable } from 'tsd';
 
 const preMiddlewareFn: PreSaveMiddlewareFunction<Document> = function(next, opts) {
   this.$markValid('name');
@@ -171,5 +171,15 @@ function gh11257() {
 
   schema.pre('updateOne', { document: false, query: true }, function() {
     this.find();
+  });
+}
+
+function gh13601() {
+  const testSchema = new Schema({
+    name: String
+  });
+
+  testSchema.pre('deleteOne', { document: true }, function() {
+    expectAssignable<Document>(this);
   });
 }

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -378,11 +378,22 @@ declare module 'mongoose' {
     pre<T = never>(method: MongooseQueryOrDocumentMiddleware | MongooseQueryOrDocumentMiddleware[] | RegExp, options: SchemaPreOptions & { document: false, query: false }, fn: PreMiddlewareFunction<T>): this;
     pre<T = never>(method: MongooseDistinctQueryMiddleware|MongooseDistinctQueryMiddleware[], options: SchemaPreOptions & { document: boolean, query: false }, fn: PreMiddlewareFunction<T>): this;
     pre<T = never>(method: MongooseDistinctDocumentMiddleware | MongooseDistinctDocumentMiddleware[] | RegExp, options: SchemaPreOptions & { document: false, query: boolean }, fn: PreMiddlewareFunction<T>): this;
+    // this = Union of Document and Query, could be called with any of them
+    pre<T = THydratedDocumentType | Query<any, any>>(
+      method: MongooseQueryAndDocumentMiddleware | MongooseQueryAndDocumentMiddleware[] | RegExp,
+      options: SchemaPreOptions & { document: true, query: true },
+      fn: PreMiddlewareFunction<T>
+    ): this;
     // this = Document
     pre<T = THydratedDocumentType>(method: 'save', fn: PreSaveMiddlewareFunction<T>): this;
     pre<T = THydratedDocumentType>(method: 'save', options: SchemaPreOptions, fn: PreSaveMiddlewareFunction<T>): this;
     pre<T = THydratedDocumentType>(method: MongooseDistinctDocumentMiddleware|MongooseDistinctDocumentMiddleware[], fn: PreMiddlewareFunction<T>): this;
     pre<T = THydratedDocumentType>(method: MongooseDistinctDocumentMiddleware|MongooseDistinctDocumentMiddleware[], options: SchemaPreOptions, fn: PreMiddlewareFunction<T>): this;
+    pre<T = THydratedDocumentType>(
+      method: MongooseQueryAndDocumentMiddleware | MongooseQueryAndDocumentMiddleware[] | RegExp,
+      options: SchemaPreOptions & { document: true },
+      fn: PreMiddlewareFunction<T>
+    ): this;
     pre<T = THydratedDocumentType>(method: MongooseQueryOrDocumentMiddleware | MongooseQueryOrDocumentMiddleware[] | RegExp, options: SchemaPreOptions & { document: true, query: false }, fn: PreMiddlewareFunction<T>): this;
     // this = Query
     pre<T = Query<any, any>>(method: MongooseDefaultQueryMiddleware|MongooseDefaultQueryMiddleware[], fn: PreMiddlewareFunction<T>): this;

